lib/documents.js

```diff
@@ -2716,6 +2716,253 @@ function urisOutputTransform(headers, data) {
+    const concurrentRequests = jobState.jobOptions.concurrentRequests;
+    if (concurrentRequests.multipleOf) {
+      if (!new Set(['forests', 'hosts']).has(String(concurrentRequests.multipleOf))) {
+        throw new Error('Invalid value for multipleOf. Value must be forests or hosts.');
+      }
+    } else {
+      concurrentRequests.multipleOf = 'forests';
+    }
+
+    if (concurrentRequests.multiplier && concurrentRequests.multiplier <= 0) {
+      throw new Error('concurrentRequests.multiplier cannot be less than one');
+    } else {
+      concurrentRequests.multiplier = 8;
+    }
+    jobState.jobOptions.concurrentRequests = concurrentRequests;
+  }
+  else {
+    jobState.jobOptions.concurrentRequests = {multipleOf:'forests', multiplier:8};
+  }
+  if(jobState.jobOptions.batchSize){
+    if(jobState.jobOptions.inputKind === 'array'){
+      throw new Error('batchSize not expected when inputKind is array.');
+    }
+    if(jobState.jobOptions.batchSize <=0){
+      jobState.jobOptions.batchSize = 250;
+    }
+  } else {
+    jobState.jobOptions.batchSize = 250;
+  }
+
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onReadAllInit))
+      .catch((err) => operation.stream.emit('error', err));
+
+  return duplexStream;
+}
+
+function onReadAllInit(output) {
+  let maxRequesters = 0;
+  const jobState = this;
+
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+  }
+  jobState.startTime = Date.now();
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onReadAllDocs(jobState,i);
+  }
+}
+
+function onReadAllDocs(jobState, readerId) {
+
+  if(jobState.error){
+    finishReader(jobState);
+    return;
+  }
+  const readBatchArray = [];
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 100;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.inputStream.read();
+        if(record === null) {
+          switch (readBatchArray.length) {
+            case 0: {
+              finishReader(jobState);
+              break a;
+            }
+            default: {
+              jobState.lastBatch = true;
+              break a;
+            }
+          }
+        }
+        readBatchArray.push(record);
+      }
+
+  if (readBatchArray.length !== 0) {
+    const batchdef = {uris: readBatchArray};
+    if(jobState.jobOptions.transform){
+      batchdef.transform = jobState.jobOptions.transform;
+    }
+    readDocs(jobState, true, batchdef,readBatchArray, readerId);
+  }
+}
+
+function readDocs(jobState, val, batchdef, readBatchArray, readerId){
+  readDocumentsImpl.call(jobState.docInstance, val, [batchdef])
```

Comment:
As I read the source code for readDocumentsImpl(), the 3rd argument should be either a simple array of uris or an object with a uris property.

As I read onReadAllDocs(), the batchdef parameter is an object with a uris property.

So, I suspect the call to readDocumentsImpl() should pass the batchdef without wrapping it in an array.

Also, when producing chunked output, it will be necessary to get a chunked stream from readDocumentsImpl()
The object case is much more important, so it is best to focus on that first.  It might even make sense to defer support for chunked streams to a future release.